### ISSUE TITLE
API for Batch Reading Data

### ISSUE THREAD

**Issue by Saad-K** [wbolster/happybase#185]
Hi,
Thanks for writing and maintaining this module, you are a life saver.
Issue: I am trying to read data from Hbase into a pandas Dataframe and then write out that dataframe as an H5 file. My data is pretty big around 5 GB. When I try to scan that data into dataframe thrift times out. Is there a way to read data in Batch instead of row by row?

Things I have tried,
I dont know the row keys. So right now I am doing a scan of the whole table and only getting the row keys using the filters `filter=b'FirstKeyOnlyFilter() AND KeyOnlyFilter()'`. Using this I create scanners of 100000 rows. And then I use those scanners to write data out to H5 files.

here is the error:

When trying to read a large data set we are continually getting timeout errors from hbase/thrift.

```
Traceback (most recent call last):
File "/usr/local/lib/python3.5/runpy.py", line 193, in _run_module_as_main
"main", mod_spec)
File "/usr/local/lib/python3.5/runpy.py", line 85, in _run_code
exec(code, run_globals)
File "/app/canvassai/main.py", line 3, in 
from canvassai.io.MQController import Worker
File "/app/canvassai/io/MQController.py", line 47, in 
worker.run()
File "/usr/local/lib/python3.5/site-packages/kombu/mixins.py", line 170, in run
for _ in self.consume(limit=None, **kwargs):
File "/usr/local/lib/python3.5/site-packages/kombu/mixins.py", line 192, in consume
conn.drain_events(timeout=safety_interval)
File "/usr/local/lib/python3.5/site-packages/kombu/connection.py", line 301, in drain_events
return self.transport.drain_events(self.connection, **kwargs)
File "/usr/local/lib/python3.5/site-packages/kombu/transport/pyamqp.py", line 103, in drain_events
return connection.drain_events(**kwargs)
File "/usr/local/lib/python3.5/site-packages/amqp/connection.py", line 464, in drain_events
return self.blocking_read(timeout)
File "/usr/local/lib/python3.5/site-packages/amqp/connection.py", line 469, in blocking_read
return self.on_inbound_frame(frame)
File "/usr/local/lib/python3.5/site-packages/amqp/method_framing.py", line 88, in on_frame
callback(channel, msg.frame_method, msg.frame_args, msg)
File "/usr/local/lib/python3.5/site-packages/amqp/connection.py", line 473, in on_inbound_method
method_sig, payload, content,
File "/usr/local/lib/python3.5/site-packages/amqp/abstract_channel.py", line 142, in dispatch_method
listener(*args)
File "/usr/local/lib/python3.5/site-packages/amqp/channel.py", line 1613, in _on_basic_deliver
fun(msg)
File "/usr/local/lib/python3.5/site-packages/kombu/messaging.py", line 623, in _receive_callback
return on_m(message) if on_m else self.receive(decoded, message)
File "/usr/local/lib/python3.5/site-packages/kombu/messaging.py", line 589, in receive
[callback(body, message) for callback in callbacks]
File "/usr/local/lib/python3.5/site-packages/kombu/messaging.py", line 589, in 
[callback(body, message) for callback in callbacks]
File "/app/canvassai/io/MQController.py", line 33, in on_message
PredictiveMaintenancePredictor(msg_details).run()
File "/app/canvassai/predictorconfig/predictor.py", line 184, in run
self.train_models()
File "/app/canvassai/predictorconfig/predictor.py", line 277, in train_models
df = self.get_prepared_data()
File "/app/canvassai/predictorconfig/predictor.py", line 268, in get_prepared_data
df = self.get_data(train=train)
File "/app/canvassai/predictorconfig/predictor.py", line 76, in get_data
df = getDataByCF(self.HbaseTableName, self.HBaseColumnFamily, HBASE_HOSTNAME, filters=self.DataFilters)
File "/app/canvassai/io/HBaseIO.py", line 54, in getDataByCF
limit=limit):
File "/usr/local/lib/python3.5/site-packages/happybase/table.py", line 394, in scan
scan_id, how_many)
File "/usr/local/lib/python3.5/site-packages/thriftpy/thrift.py", line 198, in _req
return self._recv(_api)
File "/usr/local/lib/python3.5/site-packages/thriftpy/thrift.py", line 230, in _recv
raise v
Hbase_thrift.IOError: IOError(message=b"org.apache.hadoop.hbase.client.RetriesExhaustedException: Can't get the location\n\tat org.apache.hadoop.hbase.client.RpcRetryingCallerWithReadReplicas.getRegionLocations(RpcRetryingCallerWithReadReplicas.java:309)\n\tat org.apache.hadoop.hbase.client.ScannerCallableWithReplicas.call(ScannerCallableWithReplicas.java:156)\n\tat org.apache.hadoop.hbase.client.ScannerCallableWithReplicas.call(ScannerCallableWithReplicas.java:60)\n\tat org.apache.hadoop.hbase.client.RpcRetryingCaller.callWithoutRetries(RpcRetryingCaller.java:200)\n\tat org.apache.hadoop.hbase.client.ClientScanner.call(ClientScanner.java:327)\n\tat org.apache.hadoop.hbase.client.ClientScanner.loadCache(ClientScanner.java:413)\n\tat org.apache.hadoop.hbase.client.ClientScanner.next(ClientScanner.java:371)\n\tat org.apache.hadoop.hbase.client.AbstractClientScanner.next(AbstractClientScanner.java:70)\n\tat org.apache.hadoop.hbase.thrift.ThriftServerRunner$HBaseHandler.scannerGetList(ThriftServerRunner.java:1413)\n\tat sun.reflect.GeneratedMethodAccessor16.invoke(Unknown Source)\n\tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.lang.reflect.Method.invoke(Method.java:498)\n\tat org.apache.hadoop.hbase.thrift.HbaseHandlerMetricsProxy.invoke(HbaseHandlerMetricsProxy.java:67)\n\tat com.sun.proxy.$Proxy11.scannerGetList(Unknown Source)\n\tat org.apache.hadoop.hbase.thrift.generated.Hbase$Processor$scannerGetList.getResult(Hbase.java:4789)\n\tat org.apache.hadoop.hbase.thrift.generated.Hbase$Processor$scannerGetList.getResult(Hbase.java:4773)\n\tat org.apache.thrift.ProcessFunction.process(ProcessFunction.java:39)\n\tat org.apache.thrift.TBaseProcessor.process(TBaseProcessor.java:39)\n\tat org.apache.hadoop.hbase.thrift.TBoundedThreadPoolServer$ClientConnnection.run(TBoundedThreadPoolServer.java:289)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\tat java.lang.Thread.run(Thread.java:748)\nCaused by: java.io.IOException: hconnection-0x2b8e8cbb closed\n\tat org.apache.hadoop.hbase.client.ConnectionManager$HConnectionImplementation.locateRegion(ConnectionManager.java:1156)\n\tat org.apache.hadoop.hbase.client.RpcRetryingCallerWithReadReplicas.getRegionLocations(RpcRetryingCallerWithReadReplicas.java:300)\n\t... 21 more\n")
```

**Comment by wbolster** [comment]
interesting problem, to which i have no clear answer... :)

fwiw, i think the thrift api is quite limited and mostly row oriented.

you may consider using smaller rows (which i believe is better hbase practice). perhaps i understand your data model incorrectly.

alternatively, you may try to fetch row keys using `filter=b'FirstKeyOnlyFilter() AND KeyOnlyFilter()'` a you described already, then fetch column names (assuming there are `> 1`), then fetch individual values?

it won't be pretty in any case i fear.

**Comment by Saad-K** [comment]
hey @wbolster, 
Thanks for your response. I looked at thrift API to find something that would've helped me but my thoughts were the same. 

I ended up using `filter=b'FirstKeyOnlyFilter() AND KeyOnlyFilter()'` to build a row-key batch list. And then used scan to scan items between those row-keys. 
I converted the scanned rows into dataframe and then appended the dataframe to an H5 file.

This way I am able to do a whole table scan without running out of memory or timing out. In myvcase the ideal batch-size to scan was 100,000 rows.
